File: PAT_Basic_Level_by_java/1083 是否存在相等的差/1083 是否存在相等的差.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

/**
 * 1083 是否存在相等的差
 * 
 * 题目描述：
 * 给定 N 张牌，正面数字为 1 到 N 的排列，反面数字为输入序列。
 * 计算每张牌正面数字（i+1）与反面数字的差的绝对值，找出出现次数大于1的差值，
 * 从大到小输出差值和对应的出现次数。
 * 
 * 实现原理：
 * 解题思路：
 * - 差值的可能范围为 0 到 N-1，可使用长度为 N 的数组统计各差值出现次数
 * - 遍历所有牌，计算每张牌的 |正面-反面|，对应计数加1
 * - 从大到小遍历差值数组，输出计数大于1的差值和次数
 * 
 * 算法核心逻辑：
 * - 用 int[] count = new int[N] 统计差值出现频次
 * - 差值计算公式：diff = Math.abs((i+1) - back)
 * - 从 N-1 向下遍历到 0，确保从大到小输出
 * 
 * 时间复杂度：O(N)，只需一次遍历统计，一次遍历输出
 * 空间复杂度：O(N)，计数数组占用 N 个整型空间
 */
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine().trim());
        String[] parts = br.readLine().trim().split(" ");
        
        int[] count = new int[N];
        
        // 统计每个差值出现的次数
        for (int i = 0; i < N; i++) {
            int back = Integer.parseInt(parts[i]);
            int diff = Math.abs((i + 1) - back);
            count[diff]++;
        }
        
        // 从大到小输出出现次数大于1的差值
        for (int i = N - 1; i >= 0; i--) {
            if (count[i] > 1) {
                System.out.println(i + " " + count[i]);
            }
        }
    }
}

File: PAT_Basic_Level_by_c/1083 是否存在相等的差/1083 是否存在相等的差.c
/*
 * 题目：1083 是否存在相等的差
 * 实现原理：
 *   给定一个排列，计算每个元素与其位置的差的绝对值，统计出现次数大于1的差值。
 *   算法步骤：
 *   1. 使用count数组统计每个差值出现的次数
 *   2. 遍历输入，计算|x - i|（i从1开始）
 *   3. 从大到小输出出现次数大于1的差值及其次数
 */
#include <stdio.h>
#include <stdlib.h>

int main() {
    int n;
    scanf("%d", &n);
    int count[10000] = {0};     // count[i]记录差值为i的出现次数
    for (int i = 1; i <= n; i++) {      // 位置i从1开始编号
        int x;
        scanf("%d", &x);
        int diff = abs(x - i);          // 元素值与位置的差
        count[diff]++;
    }
    for (int i = 9999; i >= 0; i--) {   // 从大到小输出
        if (count[i] > 1) {             // 只输出出现次数大于1的差值
            printf("%d %d\n", i, count[i]);
        }
    }
    return 0;
}